README
Benötigte Tools

NSIS: (Installation mit Wine)

http://prdownloads.sourceforge.net/nsis/nsis-2.46-setup.exe?download



Website/Changelog.txt
Version 1.1:

- Dokumentenklassen
  - Option 'inbcorr' fuer Fortfuehrung des Logos in der Bindekorrektur
  - Logo-Optionen verbessert
  - Diverse Fixes fuer Schriftauswahl

- Installer
  - Windows-Installer ueberarbeitet
  - UNIX Installations-Skript erweitert

- Beamer
  - Setze \title und \subtitle linksbuendig

- Prospekte (leaflet)
  - Darstellungsprobleme behoben (Text wurde nicht angezeigt)

- Briefe
  - Darstellung fuer englische Briefe verbessert

- Sonstiges
  - Dokumentation und Beispiele ueberarbeitet
  - Interne Verbesserungen

Version 1.0.3:

- Dokumentenklassen
  - Diverse Kompatibilitaetsfixes fuer `tikz` package
    - Unterstuetzung fuer `external` library
      Fixes: #0000104
    - Probleme mit Transparenz behoben
  - Diverse Kompatibilitaetsfixes fuer `listings` package
    - Spacing-Fehler bei `\lstinline` korrigiert
    - Korrigierte Behandlung fuer line numbers
  - Versalziffern fuer footnotes

- Dokumentation
  - Projekt-Website korrigiert und als hyperref-url gesetzt

- Installation
  - Fehlerhafte permissions in .deb-Skript korrigiert

Version 1.0.2:

- Allgemein
  - Verwendung von \tubslatexVersion in allen Paketen/Klassen

- Präsentation
  - Option 'lnum' hinzugefuegt
    Fixes: #0000136

- Dokumentenklassen
  - tubsfont-Definitionen ueberarbeitet.
    Fixes: #0000138, #0000139
  - Zusätzlicher Abstand fuer scrartcl maketitle ueber
    \@titlehead erzeugen anstelle von \clearpage, da dies bei neueren
    Koma-Klassen zu Leerseiten vor dem Titel führt
    Fixes: #0000141

- Briefklasse
  - Nur \pagemark redefinieren, um bessere Kompatibilität mit KOMA zu behalten.
  - Platzierung des Labels fuer Seitenzahl-Berechnung korrigiert
  - Setzte \familydefault zu \sffamily anstelle von \sffamily in \AtBeginLetter
  - Doppelte Redefinition von opening und Setzen vom uselettercount label korrigiert
  - Fixed #0000147

- Poster
  - Option 'logo' fuer Bulletin-Style poster aktiviert

Version 1.0:

- Allgemein
  - Korrektur in horizontaler Sekundaerlogopositionierung
  - Erster Absatz in gaussbox wird nicht eingerueckt
  - definierte vertikale Hoehe in gaussbox fuer vfill-Verwendung
  - Optionsbehandlung fuer 'imagefit' korrigiert
  - Benennung von eps-RGB Siegelbandlogo korrigiert
  - Bessere Unterstützung von unbekannten Papierformaten (tubstypearea)
  - Seitengröße mittel Optionen 'paperwidth' und 'paperheight' teilweise einstellbar
  - Optionsweiterleitung an bgelement korrigiert
  - Gauss-Boxen:
    - Platzierungsfehler behoben
    - imagefit-Funktionalität hinzugefügt
    - Optionen für Innenabstände
    - vertikale Positionierung korrigiert
  - Modul-Boxen:
    - Umgebung scifibox in modulebox umbenannt
    - parskip- und parindent-Wiederherstellung
    - verbesserte Implementierung
  - automatische Erkennung von individuellen Hoch-/Querformaten
  - verbesserte Behandlung unbekannter Formate

- Briefklasse
  - Verbesserte Behandlung fuer mehrere Briefe in selbem Dokument
    - Seitenzahl wird korrekt gesetzt
    - Seitenbreite auf Folgeseiten wird korrekt gesetzt
  - Unterstützung von englischer Ausgabe für Seitenzahl und Datumsangabe

- Beamer
  - Farbstufe der block-title-Hintergrundfarbe von 60 auf 40 korrigiert
  - Versionsnummer in Konsolenausgabe

- Dokumentenklassen
  - Ueberschriften, Tabellen und enumerations verwenden Versalziffern
  - Logopositionierung fuer plain-Style
  - Korrektur in twoside-Behandlung bei Titelseiten
  - Einrueckungen auf Titelseiten korrigiert
  - \titlegraphic als Alias fuer titlepicture
  - plain-Stil fuer tubsdoc wieder verfuegbar gemacht
  - Wiederherstellung von @upptertitleback und @lowertitleback fuer plain-Stil
  - \author auf CD-Titelseiten unterstützt nun \and
  - Unterstützung von \uppertitleback und \lowertitleback für backpage style
  - Titelseiten-Style für landscape
  - Umfassender Optionen- und Befehlsindex

- Plakate
  - Berechnungsfehler für Modulsystem korrigiert
  - Option 'logo' für Bulletin-Poster
  - Korrekte Schriftgrößendefinitionen angelegt
  - korrigierte Logo-Darstellung (landscape)

- Briefklasse
  - Koma-Variable 'fromuniversity' erlaubt Modifikation des Uni-Namens

- Fonts
  - \footnotesize-Deklaration für A0 korrigiert
  - mathsf und mathtt fuer Nexus definiert
  - Standard-Schnitte nutzen Mediävalziffern, Option 'lnum' für Versalziffern,
    Befehl \oldstylefigures angepasst, \textlnum eingefuehrt
  - Befehle \headline und \subheadline eingefuehrt
  - institute-Font auf \bfseries gesetzt
  - Farbeinstellungen bei CD-Schriften werden nur bei Option aktiviert

- Installation
  - Skript testet Verfuegbarkeit von kpsewhich
  - Unix Install-Script verbessert
  - Weitreichende funktionale Ueberarbeitung des Windows-Installers
  - Verbositaet fuer Windows-Installer verbessert, Logfile zu Debuggingzwecken
  - Debian-Paket verbessert (Lintian-Check fehlerfrei), Verwendung von debconf
    statt read, Test auf lokale fontmap zurueckgenommen
  - postinst und postrm-Skripte des deb-Paketes korrigiert und erweitert
  - deb-Abhängigkeiten erweitert für Kompatibilität mit Ubuntu 12.04

- Dokumentation
  - Installationshinweise für Mac OS-Installation hinzugefuegt
  - Paket-Dokumentationen erweitert
  - Erklärung zu Gauss-Layoutsystem, Plakaten und Flyern deutlich erweitert

- tubsdoc
  - Paket nun auch mit KOMA-Klassen zusammen nutzbar
  - Titelseitenstyle [plain] verfügbar
  - Koma-Font-Definitionen verbessert/korrigiert
  
- tubsstyle:
  - bglayout in bgsegment umbenannt (Alias erhalten)
  
- tubscolors:
  - Farben mit Präfix tubs~
  - Zusätzliche Farbe tubsRichBlack

- beamer:
  - \institute-Befehl zeigt ursprüngliches Verhalten und wird auf
    Titelseiten dargestellt
  - Minimalabstand von Titlepage-Text zu Segmentobergrenze

- Besipiel-Dateien:
  - tubsleaflet-Beispiel hinzugefuegt
  - Plakat-Besipiele überarbeitet

Version 0.3-beta2:

- Dokumentation
  - Überarbeitung der Kapitel 'Einleitung' und 'Installation'
  - Beginn von Indexerstellung

Allgemein
  - Überarbeitung des Gauss-Layout-Systems

report/test/test.sh
#!/bin/bash

set -e

setidx=0
testset() {
	# first item
	echo $1
	# other items
	echo ${@:2}
	setidx=`expr $setidx+1`
}

# load config file and set jobname
. $1.cfg
jobname=$1

exit 0

idx=0
back=0
for i in ${TITLESTYLE[@]}; do
	for j in ${DOCCLASS[@]}; do	
		cat $jobname.tex |
		sed "s/#TITLESTYLE#/$i/g" |
		sed "s/#DOCCLASS#/$j/g" |
		sed "s/#BACKSTYLE#/${BACKSTYLE[back]}/g" |
		pdflatex -jobname $jobname$i$j

		echo "
		\documentclass{scrartcl}
		\begin{document}
		\centering\sffamily\Huge
			$jobname\_$i\_$j
		\end{document}" | pdflatex -jobname fillpage$i$j
		catlist[idx]=fillpage$i$j.pdf
		catlist[idx+1]=$jobname$i$j.pdf
		idx=`expr $idx+2`
	done
	back=`expr $back+1`
done

echo "merging..."
pdftk ${catlist[@]} cat output $jobname.pdf

echo "cleaning..."
rm -f ${catlist[@]}
rm -f *.aux *.log

echo "done."


regression-test/run-test.sh
#!/bin/bash
#
# Arguments: Diff to commit ID
#-------------------------------------------------------------------------------

set -e

# commit id to diff to
DIFF_COMMIT=$1
CURRENT_COMMIT=$(git rev-parse HEAD)

# dir the temxf is in
TEXMF_DIR=../install/rawtree

OUT_DIR_CURRENT=out-current
OUT_DIR_BASE=out-base

TEXMF_CURRENT=texmf-current
TEXMF_BASE=texmf-base

if [ -z $1 ]; then
  echo "Run with 'run-test.sh COMMIT_ID'"
  exit
fi

# Create dirs for texmf files
mkdir -p texmf-base
mkdir -p texmf-current

# Make all new
make -BC $TEXMF_DIR/ font build-src

# Copy files to temporary texmf folder
cp -r $TEXMF_DIR/tex texmf-current/.
cp -r $TEXMF_DIR/doc texmf-current/.
cp -r $TEXMF_DIR/fonts texmf-current/.


# Checkout version to commit
git checkout $DIFF_COMMIT

# Make all new
make -BC $TEXMF_DIR/ font build-src

# Copy files to temporary texmf folder
cp -r $TEXMF_DIR/tex texmf-base/.
cp -r $TEXMF_DIR/doc texmf-base/.
cp -r $TEXMF_DIR/fonts texmf-base/.

# Create output directories
mkdir -p $OUT_DIR_CURRENT
mkdir -p $OUT_DIR_BASE

git checkout $CURRENT_COMMIT

# Generate all test files
for testfile in $(find test-documents -name *.tex); do
  TEXMFHOME=texmf-current:$(kpsewhich -var-value TEXMFHOME) pdflatex -output-directory $OUT_DIR_CURRENT $testfile;
  TEXMFHOME=texmf-base:$(kpsewhich -var-value TEXMFHOME) pdflatex -output-directory $OUT_DIR_BASE $testfile;

  ./pdfdiff.sh $OUT_DIR_CURRENT/$(basename $testfile .tex).pdf $OUT_DIR_BASE/$(basename $testfile .tex).pdf
done



install/rawtree/tubslatex_installer.sh
#! /bin/sh
#
# tubslatex installer script
#
# (c) 2013 by Enrico Jörns
#

# Halt on error
set -e

ECHO=/bin/echo

error_occured=0

exec 3>&2 # logging stream (file descriptor 3) defaults to STDERR
verbosity=2 # default to show warnings
silent_lvl=0
err_lvl=1
wrn_lvl=2
inf_lvl=3
dbg_lvl=4

#--- log functions
log_n() { log $silent_lvl "NOTE: $1"; } # Always prints
log_e() { log $err_lvl "ERROR: $1"; }
log_w() { log $wrn_lvl "WARNING: $1"; }
log_i() { log $inf_lvl "INFO: $1"; } # "info" is already a command
log_d() { log $dbg_lvl "DEBUG: $1"; }
log() {
  if [ $verbosity -ge $1 ]; then
    # Expand escaped characters, wrap at 70 chars, indent wrapped lines
    $ECHO "$2" | fold -w80 -s >&3 || true
  fi
  $ECHO "$2" | fold -w80 -s >> $logfile || true
}

check_os() {
  OS_NAME=""
  case "$(uname)" in
    "Darwin")
      OS_NAME="Darwin"
      log_d "OS found: Darwin"
      ;;
    "Linux")
      OS_NAME="Linux"
      log_d "OS found: Linux"
      ;;
    *)
      log_d "OS unknown: $(uname)"
  esac
}

#--- Checks for either wget or curl and stores result in 'download_cmd'
check_downloader() {

  download_cmd=""
  # check for wget
  status=0
  command -v wget >/dev/null 2>&1 || status=1
  if [ $status = 0 ]; then
    log_d "wget found"
    download_cmd="wget -q -O 	-"
    return 0
  else
    log_d "wget not found"
  fi

  # check for curl
  status=0
  command -v curl >/dev/null 2>&1 || status=1
  if [ $status = 0 ]; then
    log_d "curl found"
    download_cmd="curl -sS"
    return 0
  else
    log_d "curl not found"
  fi
  
  log_e "Download failed. Neither 'curl' nor 'wget' was found on your system."
  exit 1
}

#--- checks if sudo is available
check_sudo() {
  rootrun=""
  if sudo -v >/dev/null 2>&1; then
    log_d "sudo ok"
    rootrun="sudo"
  else
    log_d "sudo failed"
    # Check if user is root (might be unnecessary)
    if ! [ $(id -u) = 0 ]; then
      log_e "This script must be run as root" 1>&2
      exit 1
    fi
  fi
}

#--- shows short usage page
print_usage() {
  $ECHO "tubslatex Installer"
  $ECHO ""
  $ECHO "Syntax: sh install.sh [option] <path/to/tubslatex.tds.zip>"
  $ECHO ""
  $ECHO "where <path/to/tubslatex.tds.zip> can be both"
  $ECHO "a local and a remote file (http(s)/ftp)"
  $ECHO ""
  $ECHO "OPTIONS:"
  $ECHO "    -h    Show this help page"
  $ECHO "    -d    Show Debug output"
}

#--- Processes command line arguments
process_arguments() {
  zipurl=""

  for var in "$@"
  do
    case "$var" in
    "-d")
      verbosity=4
      log_d "verbosity on"
      ;;
    "-h")
      print_usage
      exit 0
      ;;
    *)
      zipurl=$var
      ;;
    esac
  done
}

#--- Downloads $zipurl to $unzipdir
download_tds_zip() {
  check_downloader
  $ECHO -n "Downloading zipped tds file..."
  if $download_cmd $zipurl > $unzipdir/texlive_tubs.zip 2>>$logfile; then
    $ECHO "done."
  else
    $ECHO "failed."
    log_e "Unable to download $zipurl."
    exit 1
  fi
}

#--- Checks for previous tubslatex installations
check_previous_tubslatex() {
  tubsartcl=$(kpsewhich tubsartcl.cls) || true
  if [ ! -z $tubsartcl ]; then
    log_w "Previous installation of tubslatex was found!"
    # Checks for installation directory
    case $tubsartcl in
      $(kpsewhich --var-value TEXMFLOCAL || "null")*)
        $ECHO "Installed in TEXMFLOCAL"
      ;;
      $(kpsewhich --var-value TEXMFHOME || "null")*)
        $ECHO "Installed in TEXMFHOME"
      ;;
      $(kpsewhich --var-value TEXMFDIST || "null")*)
        $ECHO "Installed in TEXMFDIST"
      ;;
      $(kpsewhich --var-value TEXMFMAIN || "null")*)
        $ECHO "Installed in TEXMFMAIN"
      ;;
      *)
      $ECHO "Installed in $($ECHO $tubsartcl | sed 's/tubsartcl.cls//g' || true)"
      ;;
    esac
    read -p "Do yout want to overwrite it? (y/n)" yesno
    case $yesno in
      y|Y)
        ;;
      *)
        $ECHO "Nothing will be done."
        exit 1
        ;;
    esac
  fi;
}

#-------------------------------------------------------------------------------

# generate temporary logfile
logfile=$(mktemp /tmp/texlive-tubs.log.XXXXXXXX)
chmod 777 $logfile

# Process user arguments
process_arguments $@

check_os
check_sudo

if [ -z $zipurl ]; then
  log_i "No explicit tubslatex archive given"
  read -p "Do you want to install the latest stable release automatically? (y/n)" yesno
  case $yesno in
    y|Y)
      zipurl=http://enricojoerns.de/tubslatex/tubslatex_latest_stable.tds.zip
      ;;
    *)
      $ECHO "Nothing will be done."
      $ECHO "Use -h option for usage information"
      exit 1
      ;;
  esac
fi

check_previous_tubslatex

# create tmpdir
unzipdir=$(mktemp -d /tmp/texlive-tubs.XXXXXXXX)
log_d "unzipdir dir is $unzipdir"

# download
if [ ! -f $zipurl ]; then
  log_d "File $zipurl not found local, trying to download"
  download_tds_zip
else
  log_d "File $zipurl found local"
  cp $zipurl $unzipdir/texlive_tubs.zip || log_e "Copying zip failed"
fi

# extract
$ECHO -n "Extracting..."
if unzip -o $unzipdir/texlive_tubs.zip -d $unzipdir >> $logfile 2>&1; then
	$ECHO "done."
else
	$ECHO "failed."
	log_e "Unable to extract zip archive $unzipdir/texlive_tubs.zip"
	rm -rf $unzipdir || true
	exit 1
fi
rm -f $unzipdir/texlive_tubs.zip


# Check if tlmgr is available
status=0
command -v tlmgr >/dev/null 2>&1 || status=1
if [ $status = 0 ]; then
  log_i "tlmgr installed, but not yet supported"
else
  log_d "tlmgr not installed"
fi


# Check if apt-get is available and allow automatic package installation
if [ "$OS_NAME" = "Linux" ]; then
status=0
command -v apt-get >/dev/null 2>&1 || status=1
if [ $status = 0 ]; then
  $ECHO "apt-get was found on your system."
  read -p "Do you want to install required dependencies now? [y/n]" yesno
  case $yesno in
    y|Y)
      $ECHO -n "Running apt-get. This may take some time..."
      if $rootrun apt-get -y install texlive-base texlive-latex-extra \
          texlive-fonts-extra latex-beamer texlive-fonts-recommended \
          texlive-lang-german texlive-extra-utils cm-super | tee $logfile; then
        $ECHO "done."
      else
        $ECHO "failed."
        error_occured=1
      fi
      ;;
    *)
      log_w "You may need to install tex packages manually!"
      ;;
  esac
else
  log_d "apt-get not found"
fi
fi


# get TEXMFLOCAL path
if texmfdir=$(kpsewhich --var-value TEXMFLOCAL); then
  log_d "TEXMFLOCAL found: $texmfdir"
else
  # TODO: handle?
  log_e "TEXMFLOCAL not found!"
  rm -rf $unzipdir || true
  exit 1
fi

if [ ! -d $unzipdir/doc ] || [ ! -d $unzipdir/fonts ] || [ ! -d $unzipdir/tex ]; then
  log_e "Did not find all required directories. Broken tds."
  rm -rf $unzipdir || true
  exit 1
else
  log_d "directories ok"
fi

$ECHO -n "Copying files to TEXMFLOCAL..."
if $rootrun cp -r $unzipdir/doc $unzipdir/fonts $unzipdir/tex $texmfdir/.; then
  $ECHO "done."
  rm -rf $unzipdir || true
else
  $ECHO "failed."
  rm -rf $unzipdir || true
  exit 1
fi

$ECHO -n "Running mktexlsr..."
if $rootrun mktexlsr >> $logfile 2>&1; then
  $ECHO "done."
else 
  $ECHO "failed. Output has been stored in"
  $ECHO "$logfile"
  $ECHO "Please include this file if you report a bug."
	error_occured=1
fi

$ECHO -n "Checking fontmap..."
if pdftexmap=`kpsewhich pdftex.map` >> $logfile 2>&1; then
  $ECHO "done."
  $ECHO "Found map file: $pdftexmap" >> $logfile
else
  $ECHO "failed. Output has been stored in"
  $ECHO "$logfile"
  $ECHO "Please include this file if you report a bug."
  exit 1
  error_occured=1
fi
    
if [ -z `$ECHO $pdftexmap | grep /var/lib` ]; then
  $ECHO ""
  $ECHO "Warning: You seem to use a local font map\n  ($pdftexmap)."
  $ECHO "Only global font maps are supported in this installation"
  read -p "Do you want to delete local font map? [y/n]" yesno
  case $yesno in
    y|Y)
      $ECHO -n "Deleting $pdftexmap..."
      $rootrun rm -f $pdftexmap
      $ECHO "done."
      ;;
    *)
      log_w "You may need to run updmap manually after install!"
      error_occured=0
      ;;
  esac
fi

# Touch 10local.cfg to prevent errors of buggy updmap
# TODO: This step might be skipped in texlive 201x
$rootrun mkdir -p /etc/texmf/updmap.d || true
$rootrun touch /etc/texmf/updmap.d/10local.cfg || true
      
# Check if getnonfreefonts is available
status=0
command -v getnonfreefonts >/dev/null 2>&1 || status=1
if [ $status = 0 ]; then
  log_d "getnonfreefonts found"
else
  $ECHO ""
  $ECHO "'getnonfreefonts' was not found on your system."
  $ECHO "Do you want automatically download and install it? [y/n]"
  read yesno
  if [ $yesno = "y" ]; then
    $ECHO -n "Downloading and installing getnonfreefonts..."
    if wget -P /tmp http://tug.org/fonts/getnonfreefonts/install-getnonfreefonts >> $logfile 2>&1; then
      $rootrun texlua /tmp/install-getnonfreefonts >> $logfile 2>&1
      rm -f /tmp/install-getnonfreefonts
    $ECHO "done."
    else
      $ECHO "failed. Output has been stored in"
      $ECHO "$logfile"
      $ECHO "Please include this file if you report a bug."
      $ECHO "-- You may need to install Arial later manually --"
    fi
  else
    $ECHO "Warning: Arial will not be installed!"
    $ECHO "-- You may need to install Arial later manually --"
  fi
fi

$ECHO -n "Running getnonfreefonts to install arial-urw..."
if $rootrun getnonfreefonts --sys --verbose arial-urw >> $logfile 2>&1; then
  $ECHO "done."
else
  $ECHO "failed. Output has been stored in"
  $ECHO "$logfile"
  $ECHO "Please include this file if you report a bug."
  $ECHO "-- You may need to install Arial later manually --"
  error_occured=1
fi

# Check for already available Mapfiles
# (Some updmap versions fail trying to enable an active map)
npsans=`updmap-sys --listmaps 2>&1 | grep NexusProSans | grep -v ! | grep -v disabled || true`
npserif=`updmap-sys --listmaps 2>&1 | grep NexusProSerif | grep -v ! | grep -v disabled || true`

# Enable Nexus maps
if [ -z "$npsans" ]; then
  $ECHO -n "Adding map file for NexusProSans..."
  if $rootrun updmap-sys --nomkmap --nohash --enable Map=NexusProSans.map >> $logfile 2>&1; then
    $ECHO "done."
  else
    $ECHO "failed.\n Output has been stored in"
    $ECHO "$logfile"
    $ECHO "Please include this file if you report a bug."
  fi
else
  $ECHO "NexusProSans already enabled, skipping."
fi
if [ -z "$npserif" ]; then
  $ECHO -n "Adding map file for NexusProSerif..."
  if $rootrun updmap-sys --nomkmap --nohash --enable Map=NexusProSerif.map >> $logfile 2>&1; then
    $ECHO "done."
  else
    $ECHO "failed.\n Output has been stored in"
    $ECHO "$logfile"
    $ECHO "Please include this file if you report a bug."
  fi
else
  $ECHO "NexusProSerif already enabled, skipping."
fi

# Run updmap to update font db
$ECHO -n "Running updmap-sys..."
if $rootrun updmap-sys >> $logfile 2>&1; then
  $ECHO "done."
else
  $ECHO "failed. Output has been stored in"
  $ECHO "$logfile"
  $ECHO "Please include this file if you report a bug."
  error_occured=1
fi

if [ error_occured = 1 ]; then
  log_w "An error occured during installation."
  $ECHO "See log $logfile for further information"
else
  $ECHO "---"
  $ECHO "tubslatex was installed successfully"
  log_d "Output has been stored in $logfile"
fi;


exit 0


install/rawtree/install.sh
#! /bin/sh
# postinst script for latex-beamer
#
# see: dh_installdeb(1)

set -e

error_occured=0

exec 3>&2 # logging stream (file descriptor 3) defaults to STDERR
verbosity=2 # default to show warnings
silent_lvl=0
err_lvl=1
wrn_lvl=2
inf_lvl=3
dbg_lvl=4

#--- log functions
log_n() { log $silent_lvl "NOTE: $1"; } # Always prints
log_e() { log $err_lvl "ERROR: $1"; }
log_w() { log $wrn_lvl "WARNING: $1"; }
log_i() { log $inf_lvl "INFO: $1"; } # "info" is already a command
log_d() { log $dbg_lvl "DEBUG: $1"; }
log() {
  if [ $verbosity -ge $1 ]; then
    # Expand escaped characters, wrap at 70 chars, indent wrapped lines
    echo "$2" | fold -w70 -s | sed '2~1s/^/  /' >&3 || true
  fi
}

#--- Checks for either wget or curl and stores result in 'download_cmd'
check_downloader() {

  download_cmd=""
  # check for wget
  status=0
  command -v wget >/dev/null 2>&1 || status=1
  if [ $status = 0 ]; then
    log_d "wget found"
    download_cmd="wget -q -O 	-"
    return 0
  else
    log_d "wget not found"
  fi

  # check for curl
  status=0
  command -v curl >/dev/null 2>&1 || status=1
  if [ $status = 0 ]; then
    log_d "curl found"
    download_cmd="curl -sS"
    return 0
  else
    log_d "curl not found"
  fi
  
  log_e "Download failed. Neither 'curl' nor 'wget' was found on your system."
  exit 1
}

# check if sudo is available
check_sudo() {
  rootrun=""
  if sudo -v >/dev/null 2>&1; then
    log_d "sudo ok"
    rootrun="sudo"
  else
    log_d "sudo failed"
    # Check if user is root (might be unnecessary)
    if ! [ $(id -u) = 0 ]; then
      log_e "This script must be run as root" 1>&2
      exit 1
    fi
  fi
}

print_usage() {
  echo "Usage: sh install.sh [-v] filename.tds.zip"
}

process_arguments() {
  zipurl=""

  for var in "$@"
  do
    case "$var" in
    "-v")
      verbosity=4
      log_d "verbosity on"
      ;;
    *)
#       zipurl=http://enricojoerns.de/tubslatex/0.3-beta3-r1190/tubslatex_0.3-beta3-r1190.tds.zip
      zipurl=$var
      ;;
    esac
  done
  
  if [ -z $zipurl ]; then
    print_usage
    exit 1;
  fi
}

#--- Downloads $zipurl to $unzipdir
download_tds_zip() {
  check_downloader
  echo -n "Downloading zipped tds file..."
  if $download_cmd $zipurl > $unzipdir/texlive_tubs.zip 2>$logfile; then
    echo "done."
  else
    echo "failed."
    log_e "Unable to download $zipurl."
    exit 1
  fi
}

#-------------------------------------------------------------------------------

process_arguments $@

check_sudo

# generate temporary logfile
logfile=$(mktemp -p /tmp texlive-tubs.log.XXXXXXXX)
chmod 777 $logfile



#-------------------------------------------------------------------------------


# create tmpdir
unzipdir=$(mktemp -d /tmp/texlive-tubs.XXXXXXXX)

# download
if [ ! -f $zipurl ]; then
  log_d "File $zipurl not found local, trying to download"
  download_tds_zip
else
  log_d "File $zipurl found local"
  cp $zipurl $unzipdir/texlive_tubs.zip || log_e "Copying zip failed"
fi

# extract
echo -n "Extracting..."
if unzip $unzipdir/texlive_tubs.zip -d $unzipdir >$logfile 2>&1; then
	echo "done."
else
	echo "failed."
	log_e "Unable to extract zip archive $unzipdir/texlive_tubs.zip"
	exit 1
fi
rm -f $unzipdir/texlive_tubs.zip

exit 0



# Check if tlmgr is available
status=0
command -v tlmgr >/dev/null 2>&1 || status=1
if [ $status = 0 ]; then
  echo "tlmgr installed, but not yet supported"
else
  log_d "tlmgr not installed"
fi


# Check if apt-get is available and allow automatic package installation
status=0
command -v apt-get >/dev/null 2>&1 || status=1
if [ $status = 0 ]; then
	echo "apt-get was found on your system."
	read -p "Do you want to automatically install dependencies? [y/n]" yesno
  case $yesno in
    y|Y)
      echo -n "Running apt-get. This may take some time..."
      if $rootrun apt-get -y install texlive-base texlive-latex-extra \
          texlive-fonts-extra latex-beamer texlive-fonts-recommended \
          texlive-lang-german texlive-extra-utils cm-super | tee $logfile; then
        echo "done."
      else
        echo "failed."
        error_occured=1
      fi
      ;;
    *)
      log_w "You may need to install tex packages manually after install!"
      ;;
  esac
else
	log_d "apt-get not found"
fi


# get TEXMFLOCAL path
if texmfdir=$(kpsewhich --var-value TEXMFLOCAL; then
  log_d "TEXMFLOCAL found: $texmfdir"
else
  # TODO: handle?
  log_e "TEXMFLOCAL not found!"
  exit 1
fi

if [ ! -d doc ] || [ ! -d fonts ] || [ ! -d tex ]; then
	log_e "Did not find all required directories"
	exit 1
else
	log_d "directories ok"
fi

echo -n "Copying files to TEXMFLOCAL..."
if cp -r $unzipdir/doc $unzipdir/fonts $unzipdir/tex $texmfdir/.; then
  echo "done."
else
  echo "failed."
  exit 1
fi

echo -n "Running mktexlsr..."
if $rootrun mktexlsr >> $logfile 2>&1; then
  echo "done."
else 
  echo "failed. Output has been stored in"
  echo "$logfile"
  echo "Please include this file if you report a bug."
	error_occured=1
fi

echo -n "Checking fontmap..."
if pdftexmap=`kpsewhich pdftex.map` >> $logfile 2>&1; then
  echo "done."
  echo "Found map file: $pdftexmap" >> $logfile
else
  echo "failed. Output has been stored in"
  echo "$logfile"
  echo "Please include this file if you report a bug."
  exit 1
  error_occured=1
fi
    
#    varlibmap=``
#    echo "varlibmap: $varlibmap"
if [ -z `echo $pdftexmap | grep /var/lib` ]; then
  echo ""
  echo "Warning: You seem to use a local font map\n  ($pdftexmap)."
  echo "Only global font maps are supported in this installation"
  read -p "Do you want to delete local font map? [y/n]" yesno
  case $yesno in
    y|Y)
      echo -n "Deleting $pdftexmap..."
      $rootrun rm -f $pdftexmap
      echo "done."
      ;;
    *)
      log_w "You may need to run updmap manually after install!"
      error_occured=0
      ;;
  esac
fi

# Touch 10local.cfg to prevent errors of buggy updmap
# TODO: This step might be skipped in texlive 201x
$rootrun mkdir -p /etc/texmf/updmap.d || true
$rootrun touch /etc/texmf/updmap.d/10local.cfg || true
      
# Check for getnonfreefonts and install if needed
# Is either located at /usr/bin/getnonfreefonts (as symlink)
# or at /usr/local/bin/getnonfreefonts
if ([ ! -x /usr/bin/getnonfreefonts ] && [ ! -x /usr/local/bin/getnonfreefonts ]); then
  echo ""
  echo "'getnonfreefonts' was not found on your system."
  echo "Do you want automatically download and install it? [y/n]"
  read yesno
  if [ $yesno = "y" ]; then
    echo -n "Downloading and installing getnonfreefonts..."
    if wget -P /tmp http://tug.org/fonts/getnonfreefonts/install-getnonfreefonts >> $logfile 2>&1; then
      $rootrun texlua /tmp/install-getnonfreefonts >> $logfile 2>&1
      rm -f /tmp/install-getnonfreefonts
    echo "done."
    else
      echo "failed. Output has been stored in"
      echo "$logfile"
      echo "Please include this file if you report a bug."
      echo "-- You may need to install Arial later manually --"
    fi
  else
    echo "Warning: Arial will not be installed!"
    echo "-- You may need to install Arial later manually --"
  fi
fi

echo -n "Running getnonfreefonts to install arial-urw..."
if $rootrun getnonfreefonts --sys --verbose arial-urw >> $logfile 2>&1; then
  echo "done."
else
  echo "failed. Output has been stored in"
  echo "$logfile"
  echo "Please include this file if you report a bug."
  echo "-- You may need to install Arial later manually --"
  error_occured=1
fi

# Check for already available Mapfiles
# (Some updmap versions fail trying to enable an active map)
npsans=`updmap-sys --listmaps 2>&1 | grep NexusProSerif | grep -v ! | grep -v disabled || true`
npserif=`updmap-sys --listmaps 2>&1 | grep NexusProSerif | grep -v ! | grep -v disabled || true`

# Enable Nexus maps
if [ -z "$npsans" ]; then
  echo -n "Adding map file for NexusProSans..."
  if $rootrun updmap-sys --nomkmap --nohash --enable Map=NexusProSans.map >> $logfile 2>&1; then
    echo "done."
  else
    echo "failed.\n Output has been stored in"
    echo "$logfile"
    echo "Please include this file if you report a bug."
  fi
else
  echo "NexusProSans already enabled, skipping."
fi
if [ -z "$npserif" ]; then
  echo -n "Adding map file for NexusProSerif..."
  if $rootrun updmap-sys --nomkmap --nohash --enable Map=NexusProSerif.map >> $logfile 2>&1; then
    echo "done."
  else
    echo "failed.\n Output has been stored in"
    echo "$logfile"
    echo "Please include this file if you report a bug."
  fi
else
  echo "NexusProSerif already enabled, skipping."
fi

# Run updmap to update font db
echo -n "Running updmap-sys..."
if $rootrun updmap-sys >> $logfile 2>&1; then
  echo "done."
else
  echo "failed. Output has been stored in"
  echo "$logfile"
  echo "Please include this file if you report a bug."
  error_occured=1
fi


# delete temp file if empty
# if [ -s $FILE ]; then
#   log_d "$logfile has data."
# else
#   log_d "$logfile is empty. Will be deleted"
#   rm -f $logfile
# fi

if [ error_occured = 1 ]; then
  log_w "An error occured during installation."
  echo "See log $logfile for further information"
else
  log_d "Output has been stored in $logfile"
fi;


exit 0
